src/kernel/scheduler.zig

```diff
@@ -0,0 +1,583 @@
+    // stack
+    return current_task.esp;
+}
+
+///
+/// Initialise the scheduler.
+///
+/// Arguments:
+///     IN/OUT allocator: *Allocator - The allocator to use when needing to allocate memory.
+///
+/// Error: std.mem.Allocator.Error
+///     OutOfMemory - There is no more memory. Any memory allocated will be freed on return.
+///
+pub fn init(allocator: *Allocator) Allocator.Error!void {
+    log.logInfo("Init schedule\n", .{});
+    defer log.logInfo("Done schedule\n", .{});
+    alloc = allocator;
+
+    // Will need to increase this as more Tasks are created
+    all_pids = try Bitmap(u32).init(32 * 32, alloc);
+
+    // Init the task list for round robin
+    tasks = TailQueue(*Task).init();
+
+    // Set up the init task to continue execution
+    current_task = try allocator.create(Task);
+    // This will allocate PID 0
+    current_task.pid = allocatePid();
+    current_task.stack = KERNEL_STACK_START;
+    // ESP will be saved on next schedule
+
+    // Create the idle task when there are no more tasks left
+    try createNewTask(idle);
+}
+
+// For testing the errdefer
+const FailingAllocator = std.testing.FailingAllocator;
+const testing_allocator = &std.testing.base_allocator_instance.allocator;
+
+fn test_fn1() void {}
+fn test_fn2() void {}
+
+///
+/// This will initial the globals but not the allocator as different allocators will be used
+/// for different tests.
+///
+fn setup_test() void {
+    // Init the pids
+    // Will fail the test if out of memory
+    all_pids = Bitmap(u32).init(32, testing_allocator) catch unreachable;
+
+    // Init the task list
+    tasks = TailQueue(*Task).init();
+}
+
+test "Task.init out of memory for task" {
+    // Set the global allocator
+    var fa = FailingAllocator.init(testing_allocator, 0);
+    alloc = &fa.allocator;
+
+    setup_test();
+
+    expectError(error.OutOfMemory, Task.init(test_fn1));
+
+    // Make sure any memory allocated is freed
+    expectEqual(fa.allocated_bytes, fa.freed_bytes);
+
+    // Make sure no PIDs were allocated
+    for (all_pids.bitmaps) |bitmap| {
+        expectEqual(bitmap, 0);
+    }
+}
+
+test "Task.init out of memory for stack" {
+    // Set the global allocator
+    var fa = FailingAllocator.init(testing_allocator, 1);
+    alloc = &fa.allocator;
+
+    setup_test();
+
+    expectError(error.OutOfMemory, Task.init(test_fn1));
+
+    // Make sure any memory allocated is freed
+    expectEqual(fa.allocated_bytes, fa.freed_bytes);
+
+    // Make sure no PIDs were allocated
+    for (all_pids.bitmaps) |bitmap| {
+        expectEqual(bitmap, 0);
+    }
+}
+
+test "Task.init expected setup" {
+    alloc = std.testing.allocator;
+
+    setup_test();
+
+    var task = try Task.init(test_fn1);
+    defer task.destroy();
+
+    // Will allocate the first PID 0
+    expectEqual(task.pid, 0);
+
+    // This looks pointless tbh as it is just copy paste from init function
```

Comment:
Just like Task.init I think this should be moved to the arch file

Reply:
This will be moved when the task is a separate file in the arch folder, leaving the scheduler file in the kernel folder